### PR TITLE
Target mode for systemd service helper

### PR DESCRIPTION
**Pull request by lamont-granquist**
Makes it so that local cookbook can bounce a remote service over
train/ssh/whatever.  Works at least for ubuntu.

Introduces a fairly rudimentary TrainHelpers class, using some
inspiration from ohbye.

### CODE REVIEW

File: lib/chef/mixin/train_helpers.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#--
+# Author:: Lamont Granquist <lamont@chef.io>
+# Copyright:: Copyright 2010-2018, Chef Software Inc.
```
Comment (tas50) [inline lib/chef/mixin/train_helpers.rb]: it's 2019

File: lib/chef/mixin/train_helpers.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    module TrainHelpers
+      def file_exist?(*args)
+        if Chef::Config.target_mode?
+          Chef.run_context.transport_connection.file(args[0]).exist?
+        else
+          File.exist?(*args)
+        end
+      end
+
+      # XXX: modifications to the StringIO won't get written back
```
Comment (tas50) [inline lib/chef/mixin/train_helpers.rb]: yard?

File: lib/chef/mixin/train_helpers.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      def file_open(*args, &block)
+        if Chef::Config.target_mode?
+          content = Chef.run_context.transport_connection.file(args[0]).content
```
Comment (tas50) [inline lib/chef/mixin/train_helpers.rb]: do we need any error handling here if the file can't be read or are we fine with just letting train blow up with its own error?

Comment (lamont-granquist) [reply lib/chef/mixin/train_helpers.rb]: doesn't matter right now, the only place we use it we check file_exist? beforehand.

that's problematic, because what should we do, and should we just fake out standard File.open errors so that code doesn't have to change?  or should we throw some other Chef::Exception error?  or should we just require consumer code to catch the error(s) that train throws?

and this whole approach is super hacky and i don't know how far it'll actually go, but i can't sit down and design the correct thing without doing exploratory surgery.

File: lib/chef/mixin/train_helpers.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#--
```
Comment (tas50) [inline lib/chef/mixin/train_helpers.rb]: ?

**Comment by tas50** [comment]
Tests?

**Comment by tas50** [comment]
Also the DCO is failing here

File: lib/chef/node_map.rb [hunk 5]
```diff
@@ -287,8 +287,6 @@ def canonical_matches?(canonical, matcher)
     # "provides" lines with identical filters sort by class name (ascending).
     #
     def compare_matchers(key, new_matcher, matcher)
-      cmp = compare_matcher_properties(new_matcher[:target_mode], matcher[:target_mode])
-      return cmp if cmp != 0
```
Comment (btm) [inline lib/chef/node_map.rb]: why did we not need this? I think I added it out of copy/pasta.

Comment (lamont-granquist) [reply lib/chef/node_map.rb]: yeah it was a copypasta error...

so what this does is establish the priority and uniqueness of two different rules.

that created problem where i think this line:

```
provides :service, target_mode: true
```

started behaving more like:

```
provides :service { true }
```

so that instead of being the least-specific, lowest-priority match it went to being the highest-priority and difficult to override.  i might not be remembering the case entirely correctly, but it definitely caused some provider_resolver tests to explode.  didn't show up until i started messing with the service provider (which is going to be where all the weird edge cases are).

now the meaning of the flag is identical to what it meant before, but the target_mode flag just means "also supports target mode".

Comment (lamont-granquist) [reply lib/chef/node_map.rb]: (actually its not uniqueness, because there is still that `matches_target_mode?` check, it is just used for sort order -- otherwise deleting those lines would have no-op'd it completely)
